### PR TITLE
Improve fish integration

### PR DESCRIPTION
**Pull request by bitraid**
The main purpose of this PR, is making the search syntax for file and cd, work as expected with fish key bindings (more details in commit message). A few other changes/improvements are also applied, each in its own commit for easier review.

The only thing that won't work (not specific to this script), is the `(inverse-)prefix-exact-match` for a query with a dir prefix (when using the `--walker-root` option). For example, `$HOME/^music`, won't display the items that start with `music` in home dir. From what I understand, this is a limitation of fzf:

https://github.com/junegunn/fzf/blob/03d6ba749647947935c237852860101f21c4782b/src/algo/algo.go#L912-L932

`trimmedLen` does not take account for the prepending path, and if I'm not mistaken, there is currently no way for the function to know about it either.

### CODE REVIEW

**Comment by bitraid** [comment]
Tests should work now. (The previous last commit disabled the function when stdin was not a tty)

**Comment by junegunn** [comment]
Thanks. It's a bit hard for me to see what problems this pull request is trying to solve, apart from a few style changes. Can you give some concrete examples, i.e. the use cases that don't work without the patch, but work with it? We could also consider adding some test cases if this deals with fish-specific problems.

https://github.com/junegunn/fzf/blob/add1aec685ffe9033bb1fdcaf0df0eae95c9cacd/test/test_go.rb#L3873

**Comment by bitraid** [comment]
Sure. Pressing `Ctrl-T` or `Alt-C` when current command line token contains:

- `'wild`


```
> 'wild- (line 1): Unexpected end of string, quotes are not balanced
set commandline 'wild
                ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
```

- `'wild'`
(the quotes are gone)

- `.mp3$` 

```
> .mp3$- (line 1): Expected a variable name after this $.
set commandline .mp3$
                    ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
- (line 1): Expected a variable name after this $.
fzf -m --query "mp3$"

```

- `!fire`

```
> !fire[: Missing argument at index 8
. = . -a ! != .
                ^
- (line 166):
      if [ "$dir" = "." -a (string sub -l 1 -- $commandline) != '.' ]
         ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
```

- `!^music`

```
> !^music[: Missing argument at index 8
. = . -a ! != .
                ^
- (line 166):
      if [ "$dir" = "." -a (string sub -l 1 -- $commandline) != '.' ]
         ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
```

- `!.mp3$`

```
> !.mp3$- (line 1): Expected a variable name after this $.
set commandline !.mp3$
                     ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
[: Missing argument at index 8
. = . -a ! != .
                ^
- (line 166):
      if [ "$dir" = "." -a (string sub -l 1 -- $commandline) != '.' ]
         ^
in function '__fzf_parse_commandline'
        called on line 1 of file -
in command substitution
        called on line 33 of file -
in function 'fzf-file-widget'
- (line 1): Expected a variable name after this $.
fzf -m --query ".mp3$"
                    ^
```

- `'wild"card'` or `wild\"card`

```
> 'wild"card'- (line 1): Unexpected end of string, quotes are not balanced
fzf -m --query "wild"card"
                         ^
in function 'fzf-file-widget'
```

**Comment by junegunn** [comment]
Thanks for the clarification! I'll review the code when I get some time.

File: shell/key-bindings.fish [hunk 1]
```diff
@@ -59,27 +57,27 @@ function fzf_key_bindings
...
           and commandline -- $result
         else
-          set -lx FZF_DEFAULT_OPTS (__fzf_defaults "" "--scheme=history --bind=ctrl-r:toggle-sort --wrap-sign '"\t"↳ ' --highlight-line $FZF_CTRL_R_OPTS +m")
-          set -lx FZF_DEFAULT_OPTS_FILE ''
-          builtin history -z | eval (__fzfcmd) --read0 --print0 -q '(commandline)' | read -lz result
+          set -l line 0
+          for i in (builtin history -z --reverse | string split0)
+            set line (math $line + 1)
+            string escape -n -- $line\t$i
+          end | string join0 | string replace -a '\n' '\n\t' | string unescape -n | eval (__fzfcmd) --tac --read0 --print0 -q '(commandline)' | string replace -r '^\d*\t' '' | read -lz result
```
Comment (junegunn) [inline shell/key-bindings.fish:80]: Have you compared the performance of the two versions? If the native fish script is as performant as the Perl version, we can just remove that and simplify the code.

Comment (bitraid) [reply shell/key-bindings.fish:80]: Yes, I have. On large history files with thousands of items, there is a slight delay (with perl it's instant). So, I think it's better to keep it (I forgot to write the reasoning in the commit message, sorry).

```
> time ./bench-hist-builtin.fish

________________________________________________________
Executed in  608.92 millis    fish           external
   usr time  506.29 millis    2.51 millis  503.79 millis
   sys time  138.58 millis    0.37 millis  138.22 millis


> time ./bench-hist-perl.fish

________________________________________________________
Executed in   75.61 millis    fish           external
   usr time   73.59 millis    2.85 millis   70.74 millis
   sys time   12.99 millis    0.00 millis   12.99 millis

```

**Comment by junegunn** [comment]
LGTM. Merging, thanks!

**Pull request merged**